File: frontend/src/api/fmea.ts
import http from './interceptor'

const api = http

/** FMEA rating bounds — mirrors backend schemas/knowledge.py RATING_MIN/MAX. */
export const RATING_MIN = 1
export const RATING_MAX = 10

/** High-RPN row highlighting thresholds (client-side display only). */
export const RPN_DANGER_THRESHOLD = 100
export const RPN_WARNING_THRESHOLD = 60

/** A persisted FMEA entry. `rpn` is computed server-side (severity*occurrence*detection). */
export interface FmeaRecord {
  id: string
  component_code: string
  function_name: string | null
  fault_code: string | null
  failure_mode: string
  effects: string | null
  cause: string | null
  severity: number
  occurrence: number
  detection: number
  rpn: number
  recommended_action: string | null
  created_at: string
  updated_at: string
}

/**
 * Create payload for an FMEA entry.
 *
 * Deliberately has NO `rpn` field: the server derives it as S*O*D and
 * ignores any client-supplied value. Ratings are integers 1-10.
 */
export interface FmeaCreate {
  component_code: string
  function_name?: string | null
  fault_code?: string | null
  failure_mode: string
  effects?: string | null
  cause?: string | null
  severity: number
  occurrence: number
  detection: number
  recommended_action?: string | null
}

/**
 * Partial-update payload for an FMEA entry.
 *
 * Every field optional; supplied ratings still constrained to 1-10. `rpn` is
 * absent — always derived server-side.
 */
export interface FmeaUpdate {
  component_code?: string
  function_name?: string | null
  fault_code?: string | null
  failure_mode?: string
  effects?: string | null
  cause?: string | null
  severity?: number
  occurrence?: number
  detection?: number
  recommended_action?: string | null
}

/** Paginated list response ({items,total}) returned by GET /fmea. */
export interface FmeaListResponse {
  items: FmeaRecord[]
  total: number
}

/** Query parameters for the list endpoint. */
export interface FmeaListParams {
  component_code?: string
  fault_code?: string
  skip?: number
  limit?: number
}

/** Compute RPN from the three ratings (client-side display; server is authoritative). */
export function computeRpn(severity: number, occurrence: number, detection: number): number {
  return severity * occurrence * detection
}

/** True when a rating is an integer within the allowed 1-10 range. */
export function isValidRating(value: number | null | undefined): boolean {
  return (
    typeof value === 'number' &&
    Number.isInteger(value) &&
    value >= RATING_MIN &&
    value <= RATING_MAX
  )
}

/** Row risk band derived from RPN, for color-coded highlighting. */
export type RpnBand = 'danger' | 'warning' | 'normal'

export function rpnBand(rpn: number): RpnBand {
  if (rpn >= RPN_DANGER_THRESHOLD) return 'danger'
  if (rpn >= RPN_WARNING_THRESHOLD) return 'warning'
  return 'normal'
}

/** List FMEA entries with optional filters. */
export async function fetchFmeas(params: FmeaListParams = {}): Promise<FmeaListResponse> {
  const query: Record<string, string | number> = {}
  if (params.component_code) query.component_code = params.component_code
  if (params.fault_code) query.fault_code = params.fault_code
  if (params.skip !== undefined) query.skip = params.skip
  if (params.limit !== undefined) query.limit = params.limit
  const response = await api.get<FmeaListResponse>('/fmea', { params: query })
  return response.data
}

/** Get a single FMEA entry by id. */
export async function fetchFmea(id: string): Promise<FmeaRecord> {
  const response = await api.get<FmeaRecord>(`/fmea/${encodeURIComponent(id)}`)
  return response.data
}

/** Create an FMEA entry. The returned record carries the server-computed rpn. */
export async function createFmea(data: FmeaCreate): Promise<FmeaRecord> {
  const response = await api.post<FmeaRecord>('/fmea', data)
  return response.data
}

/** Partially update an FMEA entry; the server recomputes rpn. */
export async function updateFmea(id: string, data: FmeaUpdate): Promise<FmeaRecord> {
  const response = await api.put<FmeaRecord>(`/fmea/${encodeURIComponent(id)}`, data)
  return response.data
}

/** Delete an FMEA entry. */
export async function deleteFmea(id: string): Promise<void> {
  await api.delete(`/fmea/${encodeURIComponent(id)}`)
}

File: src/ate_cloud/api/v1/fmea.py
"""FMEA CRUD API endpoints (task 13).

REST surface over the task-10 ``FMEA`` ORM model
(``models/knowledge.py``). RPN (risk priority number) is ALWAYS derived
server-side as severity*occurrence*detection by the model's SQLAlchemy
mapper events — the create/update schemas accept no ``rpn`` and a
client-supplied ``rpn`` is ignored. Ratings are constrained to [1, 10] at
the Pydantic boundary (-> 422), with the ORM validator and a DB CHECK
constraint as defense in depth.

Endpoints (all JWT-protected at mount time via ``_PROTECTED_ROUTERS``):
- POST   /api/v1/fmea          create an entry (201, computed rpn)
- GET    /api/v1/fmea          list, optional component_code/fault_code filters
- GET    /api/v1/fmea/{fmea_id} get one (404 when missing)
- PUT    /api/v1/fmea/{fmea_id} partial update (rpn recomputed)
- DELETE /api/v1/fmea/{fmea_id} delete (404 when missing, like products)
"""

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from ate_cloud.db import get_db
from ate_cloud.models.knowledge import FMEA
from ate_cloud.schemas.knowledge import FMEACreate, FMEAResponse, FMEAUpdate

router = APIRouter(prefix="/fmea", tags=["fmea"])

# Type alias for async DB session dependency (avoids B008 ruff warning).
DBSession = Annotated[AsyncSession, Depends(get_db)]


@router.get("", response_model=dict[str, object])
async def list_fmeas(
    db: DBSession,
    component_code: Annotated[
        str | None, Query(description="Filter by component_code")
    ] = None,
    fault_code: Annotated[
        str | None, Query(description="Filter by fault_code")
    ] = None,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=1000),
) -> dict[str, object]:
    """GET /api/v1/fmea — list FMEA entries with optional filters.

    Args:
        db: Database session.
        component_code: Optional component filter.
        fault_code: Optional fault-code filter.
        skip: Number of records to skip.
        limit: Maximum records to return.

    Returns:
        dict with ``items`` (FMEAResponse list) and ``total`` count.
    """
    filters = []
    if component_code is not None:
        filters.append(FMEA.component_code == component_code)
    if fault_code is not None:
        filters.append(FMEA.fault_code == fault_code)

    count_stmt = select(func.count()).select_from(FMEA)
    list_stmt = select(FMEA).order_by(FMEA.created_at.desc())
    for condition in filters:
        count_stmt = count_stmt.where(condition)
        list_stmt = list_stmt.where(condition)

    total = (await db.execute(count_stmt)).scalar() or 0
    rows = (await db.execute(list_stmt.offset(skip).limit(limit))).scalars().all()

    return {
        "items": [FMEAResponse.model_validate(r) for r in rows],
        "total": total,
    }


@router.get("/{fmea_id}", response_model=FMEAResponse)
async def get_fmea(fmea_id: str, db: DBSession) -> FMEAResponse:
    """GET /api/v1/fmea/{fmea_id} — fetch one entry.

    Raises:
        HTTPException: 404 if no entry exists for the id.
    """
    entry = (
        await db.execute(select(FMEA).where(FMEA.id == fmea_id))
    ).scalar_one_or_none()
    if entry is None:
        raise HTTPException(status_code=404, detail="FMEA entry not found")
    return FMEAResponse.model_validate(entry)


@router.post("", response_model=FMEAResponse, status_code=status.HTTP_201_CREATED)
async def create_fmea(data: FMEACreate, db: DBSession) -> FMEAResponse:
    """POST /api/v1/fmea — create an FMEA entry.

    The rpn is derived server-side (S*O*D) by the model's mapper events;
    any client-supplied rpn is ignored.
    """
    entry = FMEA(id=str(uuid.uuid4()), **data.model_dump())
    db.add(entry)
    await db.commit()
    await db.refresh(entry)
    return FMEAResponse.model_validate(entry)


@router.put("/{fmea_id}", response_model=FMEAResponse)
async def update_fmea(
    fmea_id: str, data: FMEAUpdate, db: DBSession
) -> FMEAResponse:
    """PUT /api/v1/fmea/{fmea_id} — partially update an entry.

    Supplied ratings are range-validated (422) and rpn is recomputed
    server-side; a client-supplied rpn is ignored.

    Raises:
        HTTPException: 404 if no entry exists for the id.
    """
    entry = (
        await db.execute(select(FMEA).where(FMEA.id == fmea_id))
    ).scalar_one_or_none()
    if entry is None:
        raise HTTPException(status_code=404, detail="FMEA entry not found")

    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(entry, key, value)

    await db.commit()
    await db.refresh(entry)
    return FMEAResponse.model_validate(entry)


@router.delete("/{fmea_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_fmea(fmea_id: str, db: DBSession) -> None:
    """DELETE /api/v1/fmea/{fmea_id} — delete an entry.

    Raises:
        HTTPException: 404 if no entry exists for the id.
    """
    entry = (
        await db.execute(select(FMEA).where(FMEA.id == fmea_id))
    ).scalar_one_or_none()
    if entry is None:
        raise HTTPException(status_code=404, detail="FMEA entry not found")

    await db.delete(entry)
    await db.commit()
